<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from diff_wheeled_robot.xacro       | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="differential_wheeled_robot" 
  xmlns:xacro="http://www.ros.org/wiki/xacro">
  <!-- Wheels -->
  <!-- in kg-->
  <!-- Defining the colors used in this robot -->
  <material name="Black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="Red">
    <color rgba="0.8 0.0 0.0 1.0"/>
  </material>
  <material name="White">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>
  <material name="Blue">
    <color rgba="0.0 0.0 0.8 1.0"/>
  </material>
  <!-- PROPERTY LIST -->
  <!--All units in m-kg-s-radians unit system -->
  <!-- Main body radius and height -->
  <!-- Main Body Cylinder base   -->
  <!-- in kg-->
  <!-- caster wheel radius and height -->
  <!-- caster wheel mass -->
  <!-- in kg-->
  <!-- caster wheel radius and height -->
  <!-- caster wheel mass -->
  <!-- in kg-->
  <!-- Wheels -->
   --&gt; 
                                    <!-- in kg-->
  <!-- Hokuyo Laser scanner -->
  <!-- Macro for calculating inertia of cylinder -->
  <!-- BASE-FOOTPRINT -->
  <!-- base_footprint is a fictitious link(frame) that is on the ground right below base_link origin -->
  <!-- <link name="base_footprint">
    <inertial>
      <mass value="0.0001"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0001" ixy="0.0" ixz="0.0" iyy="0.0001" iyz="0.0" izz="0.0001"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.001 0.001 0.001"/>
      </geometry>
    </visual>
  </link>
  <gazebo reference="base_footprint">
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <joint name="base_footprint_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.04"/>
    <parent link="base_footprint"/>
    <child link="base_link"/>
  </joint> -->
  <!-- BASE-LINK -->
  <!--Actual body/chassis of the robot-->
  <link name="base_link">
    <inertial>
      <mass value="5"/>
      <origin xyz="0 0 0"/>
      <!--The 3x3 rotational inertia matrix. -->
      <inertia ixx="0.0282916666667" ixy="0" ixz="0" iyy="0.0282916666667" iyz="0" izz="0.05625"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.2 0.2 0.05"/>
      </geometry>
      <material name="White"/>
    </visual>
    <collision>
      <origin rpy="0 0 0 " xyz="0 0 0"/>
      <geometry>
        <box size="0.28 0.17 0.05"/>
      </geometry>
    </collision>
  </link>
  <gazebo reference="base_link">
    <material>Gazebo/White</material>
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <!--fb : front, back ; lr: left, right -->
  <link name="front_right_wheel">
    <visual>
      <origin rpy="1.57079632679 0  0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.03" radius="0.032"/>
      </geometry>
      <material name="DarkGray"/>
    </visual>
    <collision>
      <origin rpy="1.57079632679 0 0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.03" radius="0.032"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="2.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.00108333333333" ixy="0" ixz="0" iyy="0.00108333333333" iyz="0" izz="0.002"/>
    </inertial>
  </link>
  <gazebo reference="front_right_wheel">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
    <material>Gazebo/Grey</material>
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <joint name="front_right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="front_right_wheel"/>
    <origin rpy="0 0 0" xyz="0.07 -0.095 0.0"/>
    <axis rpy="0  0" xyz="0 1 0"/>
    <limit effort="100" velocity="100"/>
    <joint_properties damping="0.0" friction="0.0"/>
  </joint>
  <!-- Transmission is important to link the joints and the controller -->
  <transmission name="front_right_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="front_right_wheel_joint"/>
    <actuator name="front_right_wheel_joint_motor">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!--fb : front, back ; lr: left, right -->
  <link name="front_left_wheel">
    <visual>
      <origin rpy="1.57079632679 0  0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.03" radius="0.032"/>
      </geometry>
      <material name="DarkGray"/>
    </visual>
    <collision>
      <origin rpy="1.57079632679 0 0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.03" radius="0.032"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="2.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.00108333333333" ixy="0" ixz="0" iyy="0.00108333333333" iyz="0" izz="0.002"/>
    </inertial>
  </link>
  <gazebo reference="front_left_wheel">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
    <material>Gazebo/Grey</material>
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <joint name="front_left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="front_left_wheel"/>
    <origin rpy="0 0 0" xyz="0.07 0.095 0.0"/>
    <axis rpy="0  0" xyz="0 1 0"/>
    <limit effort="100" velocity="100"/>
    <joint_properties damping="0.0" friction="0.0"/>
  </joint>
  <!-- Transmission is important to link the joints and the controller -->
  <transmission name="front_left_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="front_left_wheel_joint"/>
    <actuator name="front_left_wheel_joint_motor">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="back_right_wheel">
    <visual>
      <origin rpy="1.57079632679 0  0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.03" radius="0.032"/>
      </geometry>
      <material name="DarkGray"/>
    </visual>
    <collision>
      <origin rpy="1.57079632679 0 0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.03" radius="0.032"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="2.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.00108333333333" ixy="0" ixz="0" iyy="0.00108333333333" iyz="0" izz="0.002"/>
    </inertial>
  </link>
  <gazebo reference="back_right_wheel">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
    <material>Gazebo/Grey</material>
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <joint name="back_right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="back_right_wheel"/>
    <origin rpy="0 0 0" xyz="-0.07 -0.095 0.0"/>
    <axis rpy="0  0" xyz="0 1 0"/>
    <limit effort="100" velocity="100"/>
    <joint_properties damping="0.0" friction="0.0"/>
  </joint>
  <!-- Transmission is important to link the joints and the controller -->
  <transmission name="back_right_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="back_right_wheel_joint"/>
    <actuator name="back_right_wheel_joint_motor">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!--fb : front, back ; lr: left, right -->
  <link name="back_left_wheel">
    <visual>
      <origin rpy="1.57079632679 0  0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.03" radius="0.032"/>
      </geometry>
      <material name="DarkGray"/>
    </visual>
    <collision>
      <origin rpy="1.57079632679 0 0 " xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.03" radius="0.032"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="2.5"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.00108333333333" ixy="0" ixz="0" iyy="0.00108333333333" iyz="0" izz="0.002"/>
    </inertial>
  </link>
  <gazebo reference="back_left_wheel">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
    <material>Gazebo/Grey</material>
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <joint name="back_left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="back_left_wheel"/>
    <origin rpy="0 0 0" xyz="-0.07 0.095 0.0"/>
    <axis rpy="0  0" xyz="0 1 0"/>
    <limit effort="100" velocity="100"/>
    <joint_properties damping="0.0" friction="0.0"/>
  </joint>
  <!-- Transmission is important to link the joints and the controller -->
  <transmission name="back_left_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="back_left_wheel_joint"/>
    <actuator name="back_left_wheel_joint_motor">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!-- SENSORS -->
  <!-- camera -->
  <link name="camera_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Blue"/>
    </visual>
  </link>
  <joint name="camera_joint" type="fixed">
    <origin rpy="0 0 -1.57" xyz="0 0 0.1"/>
    <parent link="base_link"/>
    <child link="camera_link"/>
  </joint>
  <link name="imu_link_">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Black"/>
    </visual>
  </link>
  <joint name="imu_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.05"/>
    <parent link="base_link"/>
    <child link="imu_link_"/>
  </joint>
  <!-- Hokuyo Laser -->
  <joint name="hokuyo_joint" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="0.065 0 0.0325" rpy="0 0 -3.14"/>
    <parent link="base_link"/>
    <child link="hokuyo_link"/>
  </joint>

  <link name="hokuyo_link">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size=".05 .05 .05"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://gx2019_omni_simulations/urdf/meshes/hokuyo.dae"/>
      </geometry>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>
  <gazebo reference="camera_link">
    <material>Gazebo/Green</material>
    <sensor type="camera" name="camera1">
      <update_rate>30.0</update_rate>
      <camera name="head">
        <horizontal_fov>1.3962634</horizontal_fov>
        <image>
          <width>800</width>
          <height>800</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.02</near>
          <far>300</far>
        </clip>
      </camera>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <alwaysOn>true</alwaysOn>
        <updateRate>0.0</updateRate>
        <cameraName>my_robot/camera1</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <frameName>camera</frameName>
        <hackBaseline>0.07</hackBaseline>
        <distortionK1>0.0</distortionK1>
        <distortionK2>0.0</distortionK2>
        <distortionK3>0.0</distortionK3>
        <distortionT1>0.0</distortionT1>
        <distortionT2>0.0</distortionT2>
      </plugin>
    </sensor>
  </gazebo>

  <!-- hokuyo laser sensor-->
  <gazebo reference="hokuyo_link">
    <sensor type="gpu_ray" name="head_hokuyo_sensor">
      <pose>0 0 0 0 0 0</pose>
      <visualize>false</visualize>
      <update_rate>40</update_rate>
      <ray>
        <scan>
          <horizontal>
            <samples>360</samples>
            <resolution>1</resolution>
            <min_angle>-3.1415926</min_angle>
            <max_angle>3.1415926</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.01</min>
          <max>10.0</max>
          <resolution>0.01</resolution>
        </range>
        <noise>
          <type>gaussian</type>
          <!-- Noise parameters based on published spec for Hokuyo laser
               achieving "+-30mm" accuracy at range < 10m.  A mean of 0.0m and
               stddev of 0.01m will put 99.7% of samples within 0.03m of the true
               reading. -->
          <mean>0.0</mean>
          <stddev>0.01</stddev>
        </noise>
      </ray>
      <plugin name="gazebo_ros_head_hokuyo_controller" filename="libgazebo_ros_gpu_laser.so">
        <topicName>/my_robot/laser/scan</topicName>
        <frameName>hokuyo_link</frameName>
      </plugin>
    </sensor>
  </gazebo>


  <!-- Differential drive controller  -->
  <!-- <gazebo>
    <plugin filename="libgazebo_ros_diff_drive.so" name="differential_drive_controller_front">
      <rosDebugLevel>Debug</rosDebugLevel>
      <publishWheelTF>True</publishWheelTF>
      <publishTf>1</publishTf>
      <publishWheelJointState>true</publishWheelJointState>
      <alwaysOn>true</alwaysOn>
      <updateRate>100.0</updateRate>
      <leftJoint>front_left_wheel_joint</leftJoint>
      <rightJoint>front_right_wheel_joint</rightJoint>
      <wheelSeparation>0.3</wheelSeparation>
      <wheelDiameter>0.05</wheelDiameter>
      <broadcastTF>1</broadcastTF>
      <wheelTorque>30</wheelTorque>
      <wheelAcceleration>1.8</wheelAcceleration>
      <commandTopic>cmd_vel</commandTopic>
      <odometrySource>odom</odometrySource>    
      <publishTf>1</publishTf>
      <odometryFrame>odom</odometryFrame>
      <odometryTopic>odom</odometryTopic>
      <robotBaseFrame>base_link</robotBaseFrame>
    </plugin>
  </gazebo> -->
  <gazebo>
    <plugin name="object_controller" filename="libgazebo_ros_planar_move.so">
      <commandTopic>cmd_vel</commandTopic>
      <!-- <odometryRate>20.0</odometryRate> -->
      <broadcastTF>false</broadcastTF>
      <publishOdomTF>false</publishOdomTF>
      <robotBaseFrame>base_link</robotBaseFrame>
    </plugin>
  </gazebo>
</robot>

